File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();

    void fun(int i, int tar, int[] a, List<Integer> list) {
        if (tar == 0) {
            ans.add(new ArrayList<>(list));
            return;
        }
        if (tar < 0 || i == a.length) {
            return;
        }
        list.add(a[i]);
        fun(i, tar - a[i], a, list);
        list.remove(list.size() - 1);
        fun(i + 1, tar, a, list);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        fun(0, target, candidates, new ArrayList<>());
        return ans;
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def __init__(self):
        self.ans = []

    def fun(self, i, tar, a, lst):
        if tar == 0:
            self.ans.append(lst[:])
            return
        if tar < 0 or i == len(a):
            return
        lst.append(a[i])
        self.fun(i, tar - a[i], a, lst)
        lst.pop()
        self.fun(i + 1, tar, a, lst)

    def combinationSum(self, candidates, target):
        self.fun(0, target, candidates, [])
        return self.ans
